File: marvelHeroes/src/main/java/com/futureworkshops/marvelheroes/domain/repositories/character/CharacterQuery.java
/*
 * Copyright (c) 2018 FutureWorkshops. All rights reserved.
 */

package com.futureworkshops.marvelheroes.domain.repositories.character;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Utility class used for creating queries that match the Marvel API.
 *
 * <p>Inspired from <a href="https://github.com/Karumi/MarvelApiClientAndroid/blob/master/MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/java">Marvel API Client repository</a></p>
 *
 * Created by stelian on 04/04/2018.
 */

public class CharacterQuery {
    
    private static final String QUERY_NAME = "name";
    private static final String QUERY_NAME_START_WITH = "nameStartsWith";
    private static final String QUERY_MODIFIED_SINCE = "modifiedSince";
    private static final String QUERY_COMICS = "comics";
    private static final String QUERY_SERIES = "series";
    private static final String QUERY_EVENTS = "events";
    private static final String QUERY_STORIES = "stories";
    private static final String QUERY_ORDER_BY = "orderBy";
    private static final String QUERY_LIMIT = "limit";
    private static final String QUERY_OFFSET = "offset";
    
    private String name;
    private String nameStartWith;
    private String modifiedSince;
    private String comics;
    private String series;
    private String events;
    private String stories;
    private String orderBy;
    private int limit;
    private int offset;
    
    public CharacterQuery(String name, String nameStartWith, String modifiedSince, String comics,
                          String series, String events, String stories,
                          String orderBy, int limit, int offset) {
        this.name = name;
        this.nameStartWith = nameStartWith;
        this.modifiedSince = modifiedSince;
        this.comics = comics;
        this.series = series;
        this.events = events;
        this.stories = stories;
        this.orderBy = orderBy;
        this.limit = limit;
        this.offset = offset;
    }
    
    
    public Map<String, Object> toMap() {
        Map<String, Object> returnValues = new HashMap<>();
        
        if (name != null) {
            returnValues.put(QUERY_NAME, name);
        }
        
        if (nameStartWith != null) {
            returnValues.put(QUERY_NAME_START_WITH, nameStartWith);
        }
        
        if (modifiedSince != null) {
            returnValues.put(QUERY_MODIFIED_SINCE, modifiedSince);
        }
        
        if (comics != null) {
            returnValues.put(QUERY_COMICS, comics);
        }
        
        if (series != null) {
            returnValues.put(QUERY_SERIES, series);
        }
        
        if (events != null) {
            returnValues.put(QUERY_EVENTS, events);
        }
        
        if (stories != null) {
            returnValues.put(QUERY_STORIES, stories);
        }
        
        if (orderBy != null) {
            returnValues.put(QUERY_ORDER_BY, orderBy);
        }
        
        if (limit > 0) {
            returnValues.put(QUERY_LIMIT, limit);
        }
        
        if (offset > 0) {
            returnValues.put(QUERY_OFFSET, offset);
        }
        
        return returnValues;
    }
    
    /**
     * Builder used for simplifying the Character querying process.
     */
    public static final class Builder {
        
        public final static int MAX_SIZE = 100;
        public static final String LIST_SEPARATOR = ",";
        
        private static final String DATE_PATTERN = "yyyy-MM-dd'T'HH:mm:ssZ";
        private static final String ORDER_BY_NAME = "name";
        private static final String ORDER_BY_MODIFIED = "modified";
        
        private String name;
        private String nameStartWith;
        private Date modifiedSince;
        private List<Integer> comics = new ArrayList<>();
        private List<Integer> series = new ArrayList<>();
        private List<Integer> events = new ArrayList<>();
        private List<Integer> stories = new ArrayList<>();
        private String orderBy;
        private boolean orderAsscending = false;
        private int limit;
        private int offset;
        
        
        public Builder() {
        }
        
        public Builder name(String val) {
            name = val;
            return this;
        }
        
        public Builder nameStartWith(String val) {
            nameStartWith = val;
            return this;
        }
        
        public Builder modifiedSince(Date val) {
            modifiedSince = val;
            return this;
        }
        
        public Builder comic(int comic) {
            comics.add(comic);
            return this;
        }
        
        public Builder comics(List<Integer> comics) {
            checkNull(comics);
            this.comics.addAll(comics);
            return this;
        }
        
        public Builder serie(int serie) {
            series.add(serie);
            return this;
        }
        
        public Builder series(List<Integer> series) {
            checkNull(series);
            this.series.addAll(series);
            return this;
        }
        
        public Builder event(int event) {
            events.add(event);
            return this;
        }
        
        public Builder events(List<Integer> events) {
            checkNull(events);
            this.events.addAll(events);
            return this;
        }
        
        public Builder story(int story) {
            stories.add(story);
            return this;
        }
        
        public Builder stories(List<Integer> stories) {
            checkNull(stories);
            this.stories.addAll(stories);
            return this;
        }
        
        public Builder orderByName(boolean asscending) {
            orderBy = ORDER_BY_NAME;
            orderAsscending = asscending;
            return this;
        }
        
        public Builder orderByLastModified(boolean asscending) {
            orderBy = ORDER_BY_MODIFIED;
            orderAsscending = asscending;
            return this;
        }
        
        public Builder limit(int val) {
            checkLimit(val);
            limit = val;
            return this;
        }
        
        public Builder offset(int val) {
            if (val < 0) {
                throw new IllegalArgumentException("offset must be bigger or equals than zero");
            }
            
            this.offset = val;
            return this;
        }
        
        public CharacterQuery build() {
            String plainModifedSince = convertDate(modifiedSince);
            String plainComics = convertToList(comics);
            String plainEvents = convertToList(events);
            String plainSeries = convertToList(series);
            String plainStories = convertToList(stories);
            String plainOrderBy = convertOrderBy(orderBy, orderAsscending);
            
            return new CharacterQuery(this.name, this.nameStartWith, plainModifedSince,
                plainComics, plainSeries, plainEvents, plainStories, plainOrderBy,
                this.limit, this.offset);
        }
        
        private String convertDate(Date date) {
            if (date == null) {
                return null;
            }
            
            SimpleDateFormat format = new SimpleDateFormat(DATE_PATTERN);
            return format.format(date);
        }
        
        private void checkLimit(int limit) {
            if (limit <= 0) {
                throw new IllegalArgumentException("limit must be bigger than zero");
            }
            
            if (limit > MAX_SIZE) {
                throw new IllegalArgumentException("limit must be smaller than 100");
            }
        }
        
        private void checkNull(List<Integer> list) {
            if (list == null) {
                throw new IllegalArgumentException("the collection can not be null");
            }
        }
        
        private String convertToList(List<Integer> list) {
            String plainList = "";
            for (int i = 0; i < list.size(); i++) {
                plainList += Integer.toString(list.get(i));
                if (i < list.size() - 1) {
                    plainList += LIST_SEPARATOR;
                }
            }
            return (plainList.isEmpty()) ? null : plainList;
        }
        
        private String convertOrderBy(String orderBy, boolean ascendant) {
            if (orderBy == null) {
                return null;
            }
            
            String plainOrderBy = orderBy.toString();
            return (ascendant) ? plainOrderBy : "-" + plainOrderBy;
        }
    }
}

File: marvelHeroes/src/main/java/com/futureworkshops/marvelheroes/domain/repositories/character/CharacterQuery.kt
/*
 * Copyright (c) 2018 FutureWorkshops. All rights reserved.
 */

package com.futureworkshops.marvelheroes.domain.repositories.character

import java.text.SimpleDateFormat
import java.util.*

/**
 * Utility class used for creating queries that match the Marvel API.
 *
 *
 * Inspired from [Marvel API Client repository](https://github.com/Karumi/MarvelApiClientAndroid/blob/master/MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/java)
 *
 * Created by stelian on 04/04/2018.
 */

class CharacterQuery(private val name: String?, private val nameStartWith: String?, private val modifiedSince: String?, private val comics: String?,
                     private val series: String?, private val events: String?, private val stories: String?,
                     private val orderBy: String?, private val limit: Int, private val offset: Int) {
    
    
    fun toMap(): Map<String, Any> {
        return HashMap<String, Any>()
                .apply {
                    name?.let { this[QUERY_NAME] = it }
                    nameStartWith?.let { this[QUERY_NAME_START_WITH] = it }
                    modifiedSince?.let { this[QUERY_MODIFIED_SINCE] = it }
                    comics?.let { this[QUERY_COMICS] = it }
                    series?.let { this[QUERY_SERIES] = it }
                    events?.let { this[QUERY_EVENTS] = it }
                    stories?.let { this[QUERY_STORIES] = it }
                    orderBy?.let { this[QUERY_ORDER_BY] = it }
                    limit.takeIf { it > 0 }?.let { this[QUERY_LIMIT] = limit }
                    offset.takeIf { it > 0 }?.let { this[QUERY_OFFSET] = offset }
                }
    }
    
    
    /**
     * Builder used for simplifying the MarvelCharacter querying process.
     */
    class Builder(var name: String? = null,
                  var nameStartWith: String? = null,
                  var modifiedSince: Date? = null,
                  var comics: MutableList<Int> = mutableListOf(),
                  var series: MutableList<Int> = mutableListOf(),
                  var events: MutableList<Int> = mutableListOf(),
                  var stories: MutableList<Int> = mutableListOf(),
                  var orderBy: String? = null,
                  var orderAscending: Boolean = false,
                  var limit: Int = 0,
                  var offset: Int = 0) {
        
        fun name(name: String): Builder {
            this.name = name
            return this
        }
        
        fun nameStartWith(nameStartWith: String): Builder {
            this.nameStartWith = nameStartWith
            return this
        }
        
        fun modifiedSince(modifiedSince: Date): Builder {
            this.modifiedSince = modifiedSince
            return this
        }
        
        fun comic(comic: Int): Builder {
            comics.add(comic)
            return this
        }
        
        fun comics(comics: List<Int>): Builder {
            checkNull(comics)
            this.comics.addAll(comics)
            return this
        }
        
        fun serie(serie: Int): Builder {
            series.add(serie)
            return this
        }
        
        fun series(series: List<Int>): Builder {
            checkNull(series)
            this.series.addAll(series)
            return this
        }
        
        fun event(event: Int): Builder {
            events.add(event)
            return this
        }
        
        fun events(events: List<Int>): Builder {
            checkNull(events)
            this.events.addAll(events)
            return this
        }
        
        fun story(story: Int): Builder {
            stories.add(story)
            return this
        }
        
        fun stories(stories: List<Int>): Builder {
            checkNull(stories)
            this.stories.addAll(stories)
            return this
        }
        
        fun orderByName(asscending: Boolean): Builder {
            orderBy = ORDER_BY_NAME
            orderAscending = asscending
            return this
        }
        
        fun orderByLastModified(asscending: Boolean): Builder {
            orderBy = ORDER_BY_MODIFIED
            orderAscending = asscending
            return this
        }
        
        fun limit(limit: Int): Builder {
            checkLimit(limit)
            this.limit = limit
            return this
        }
        
        fun offset(offset: Int): Builder {
            if (offset < 0) {
                throw IllegalArgumentException("offset must be bigger or equals than zero")
            }
            
            this.offset = offset
            return this
        }
        
        fun build(): CharacterQuery {
            val plainModifedSince = convertDate(modifiedSince)
            val plainComics = convertToList(comics)
            val plainEvents = convertToList(events)
            val plainSeries = convertToList(series)
            val plainStories = convertToList(stories)
            val plainOrderBy = convertOrderBy(orderBy, orderAscending)
            
            return CharacterQuery(this.name, this.nameStartWith, plainModifedSince,
                    plainComics, plainSeries, plainEvents, plainStories, plainOrderBy,
                    this.limit, this.offset)
        }
        
        private fun convertDate(date: Date?): String? {
            if (date == null) {
                return null
            }
            
            val format = SimpleDateFormat(DATE_PATTERN)
            return format.format(date)
        }
        
        private fun checkLimit(limit: Int) {
            if (limit <= 0) {
                throw IllegalArgumentException("limit must be bigger than zero")
            }
            
            if (limit > MAX_SIZE) {
                throw IllegalArgumentException("limit must be smaller than 100")
            }
        }
        
        private fun checkNull(list: List<Int>?) {
            if (list == null) {
                throw IllegalArgumentException("the collection can not be null")
            }
        }
        
        private fun convertToList(list: List<Int>): String? {
            var plainList = ""
            for (i in list.indices) {
                plainList += Integer.toString(list[i])
                if (i < list.size - 1) {
                    plainList += LIST_SEPARATOR
                }
            }
            return if (plainList.isEmpty()) null else plainList
        }
        
        private fun convertOrderBy(orderBy: String?, ascendant: Boolean): String? {
            if (orderBy == null) {
                return null
            }
            
            val plainOrderBy = orderBy.toString()
            return if (ascendant) plainOrderBy else "-$plainOrderBy"
        }
        
        companion object {
            
            val MAX_SIZE = 100
            val LIST_SEPARATOR = ","
            
            private val DATE_PATTERN = "yyyy-MM-dd'T'HH:mm:ssZ"
            private val ORDER_BY_NAME = "name"
            private val ORDER_BY_MODIFIED = "modified"
        }
    }
    
    companion object {
        
        private val QUERY_NAME = "name"
        private val QUERY_NAME_START_WITH = "nameStartsWith"
        private val QUERY_MODIFIED_SINCE = "modifiedSince"
        private val QUERY_COMICS = "comics"
        private val QUERY_SERIES = "series"
        private val QUERY_EVENTS = "events"
        private val QUERY_STORIES = "stories"
        private val QUERY_ORDER_BY = "orderBy"
        private val QUERY_LIMIT = "limit"
        private val QUERY_OFFSET = "offset"
    }
}
